### PR TITLE
pitcherSS

### PR DESCRIPTION
**Pull request by gsyoyo07**

### CODE REVIEW

File: src/main/java/frc/robot/constants/RobotConstants.java [hunk 1]
```diff
@@ -114,5 +114,14 @@
...
         public static double STALL_CURRENT_LIMIT = 20;
     }
+
+    public static class PitcherConstants {
+        public static final TrigonTalonSRX MOTOR = PitcherComponents.MOTOR;
+        public static final double GEAR_RATIO = 10;
+        public static final double STALL_CURRENT_LIMIT = 20;
+        public static double OPEN_ANGLE = 50;
+        public static double CLOSED_ANGLE = 0;
+        //TODO: TO SET ANGLES
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/RobotConstants.java:124]: do now

File: src/main/java/frc/robot/constants/RobotConstants.java [hunk 2]
```diff
@@ -114,5 +114,14 @@
         public static double CLOSED_ANGLE = 0;
         public static double STALL_CURRENT_LIMIT = 20;
     }
+
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/RobotConstants.java:117]: i should have told you in last commit remove closed angle

File: src/main/java/frc/robot/constants/RobotMap.java [hunk 3]
```diff
@@ -32,6 +32,7 @@
 
         public static class MOTOR_ID {
             public static final int INTAKE_OPENER_MOTOR_ID = 3;
+            public static final int PITCHER_MOTOR_ID = 12;
         }
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/RobotMap.java:36]: split into two classes

File: src/main/java/frc/robot/subsystems/Pitcher/PitcherSS.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
+package frc.robot.subsystems.Pitcher;
+
+import com.ctre.phoenix.motorcontrol.ControlMode;
+import edu.wpi.first.math.MathUtil;
+import frc.robot.components.TrigonTalonSRX;
+import frc.robot.constants.RobotConstants.PitcherConstants;
+import frc.robot.subsystems.OverridableSubsystem;
+import frc.robot.utilities.EncoderConversions;
+
+public class PitcherSS extends OverridableSubsystem {
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/Pitcher/PitcherSS.java:10]: testable not overridable

File: src/main/java/frc/robot/subsystems/Pitcher/PitcherSS.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    public PitcherSS() {
+        motor = PitcherConstants.MOTOR;
+        resetEncoder();
+    }
+
+    /**
+     * @param power to be applied to the motors
+     */
+    @Override
+    public void overriddenMove(double power) {
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/Pitcher/PitcherSS.java:22]: not overridable subsytem

File: src/main/java/frc/robot/subsystems/Pitcher/PitcherSS.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    /**
+     * Sets the intake to the given angle
+     *
+     * @param degree desired angle in degrees
+     */
+    public void setPosition(double degree) {
+        motor.set(ControlMode.Position, EncoderConversions.degreesToMag(
+                MathUtil.clamp(degree, PitcherConstants.CLOSED_ANGLE, PitcherConstants.OPEN_ANGLE),
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/Pitcher/PitcherSS.java:42]: add this fail safe to the intake opener

**Review by Mboesn — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/constants/RobotMap.java (hunk 3 above)
Comment (gsyoyo07) [reply src/main/java/frc/robot/constants/RobotMap.java:36]: ok

File: src/main/java/frc/robot/subsystems/Pitcher/PitcherSS.java (hunk 4 above)
Comment (gsyoyo07) [reply src/main/java/frc/robot/subsystems/Pitcher/PitcherSS.java:10]: ok

File: src/main/java/frc/robot/subsystems/Pitcher/PitcherSS.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    /**
+     * Sets the intake to the given angle
+     *
+     * @param degree desired angle in degrees
+     */
+    public void setAngle(double degree) {
+        degree = EncoderConversions.degreesToMag(
+                MathUtil.clamp(degree, PitcherConstants.CLOSED_ANGLE, PitcherConstants.OPEN_ANGLE),
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/Pitcher/PitcherSS.java:42]: flip PitcherConstants.CLOSED_ANGLE and PitcherConstants.OPEN_ANGLE

File: src/main/java/frc/robot/utilities/EncoderConversions.java [hunk 8]
```diff
@@ -70,11 +70,19 @@ public static double MPSToFalcon(double velocity, double circumference, double g
...
+    public static double magToDegrees(double magTick) {
+        return magToDegrees(magTick, 1);
+    }
+
+    public static double degreesToMag(double degrees, double gearRatio) {
+        return degrees * 4096f / 360 * gearRatio;
+    }
+
+    public static double degreesToMag(double degrees) {
+        return degreesToMag(degrees, 1);
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/EncoderConversions.java:86]: documentation

**Review by Mboesn — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/constants/RobotComponents.java [hunk 9]
```diff
@@ -112,6 +112,16 @@
                 withClosedLoopRampRate(0.5).
                 withPID(RobotConstants.LOCAL_CONSTANTS.localIntakeOpenerConstants.pidfCoefs);
         public static PIDFTalonSRX MOTOR = new PIDFTalonSRX(
-                CAN.MOTOR_ID.INTAKE_OPENER_MOTOR_ID, INTAKE_OPENER_MOTOR_CONFIG);
+                CAN.IntakeOpener.INTAKE_OPENER_MOTOR_ID, INTAKE_OPENER_MOTOR_CONFIG);
+    }
+
+    protected static class PitcherComponents {
+        private static final MotorConfig PITCHER_MOTOR_CONFIG = new MotorConfig().
```
Comment (ArielSegal1) [inline src/main/java/frc/robot/constants/RobotComponents.java:119]: Remove PITCHER_ in line 119.

File: src/main/java/frc/robot/constants/RobotMap.java [hunk 10]
```diff
@@ -30,9 +30,13 @@
             public static final int REAR_RIGHT_ANGLE_ENCODER_ID = 10;
         }
 
-        public static class MOTOR_ID {
+        public static class IntakeOpener {
             public static final int INTAKE_OPENER_MOTOR_ID = 3;
```
Comment (ArielSegal1) [inline src/main/java/frc/robot/constants/RobotMap.java:34]: delete INTAKE_ in line 34

File: src/main/java/frc/robot/constants/RobotMap.java [hunk 11]
```diff
@@ -30,9 +30,13 @@
...
         }
+
+        public static class Pitcher {
+            public static final int PITCHER_MOTOR_ID = 12;
```
Comment (ArielSegal1) [inline src/main/java/frc/robot/constants/RobotMap.java:38]: delete PITCHER_ in line 38

**Review by ArielSegal1 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/constants/RobotComponents.java (hunk 9 above)
Comment (gsyoyo07) [reply src/main/java/frc/robot/constants/RobotComponents.java:119]: but in intake opener is writine like that so sould we chance intake also ?

File: src/main/java/frc/robot/constants/RobotMap.java (hunk 10 above)
Comment (gsyoyo07) [reply src/main/java/frc/robot/constants/RobotMap.java:34]: @Mboesn sould we do that in the intake branch?

File: src/main/java/frc/robot/constants/RobotMap.java (hunk 11 above)
Comment (gsyoyo07) [reply src/main/java/frc/robot/constants/RobotMap.java:38]: but then we wont know were it comes from and the we would need to change the other ids

Comment (Mboesn) [reply src/main/java/frc/robot/constants/RobotMap.java:38]: Anything in pitcher belongs to the pitcher

File: src/main/java/frc/robot/constants/RobotMap.java (hunk 10 above)
Comment (Mboesn) [reply src/main/java/frc/robot/constants/RobotMap.java:34]: In theory yes, but just change it here

File: src/main/java/frc/robot/constants/RobotComponents.java (hunk 9 above)
Comment (Mboesn) [reply src/main/java/frc/robot/constants/RobotComponents.java:119]: Yes

File: src/main/java/frc/robot/constants/RobotMap.java (hunk 11 above)
Comment (gsyoyo07) [reply src/main/java/frc/robot/constants/RobotMap.java:38]: ok

File: src/main/java/frc/robot/constants/RobotMap.java (hunk 10 above)
Comment (gsyoyo07) [reply src/main/java/frc/robot/constants/RobotMap.java:34]: kkk

File: src/main/java/frc/robot/constants/RobotComponents.java (hunk 9 above)
Comment (gsyoyo07) [reply src/main/java/frc/robot/constants/RobotComponents.java:119]: ok

File: src/main/java/frc/robot/utilities/EncoderConversions.java (hunk 8 above)
Comment (gsyoyo07) [reply src/main/java/frc/robot/utilities/EncoderConversions.java:86]: ok

File: src/main/java/frc/robot/utilities/EncoderConversions.java [hunk 12]
```diff
@@ -78,6 +78,10 @@ public static double magToDegrees(double magTick) {
         return magToDegrees(magTick, 1);
     }
 
+    /**
+     * @param degrees the angle wanted to convert to ticks
+     * @return the angle as ticks
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/EncoderConversions.java:83]: @ param degrees angle of the wheel or gear in degrees
@ param gearRatio  Gear Ratio between Mag and Mechanism (set to 1 for Mag RPM)
@ return the angle of the Mag in ticks

File: src/main/java/frc/robot/constants/RobotComponents.java [hunk 13]
```diff
@@ -105,13 +105,23 @@
...
                 withPID(RobotConstants.LOCAL_CONSTANTS.localIntakeOpenerConstants.pidfCoefs);
         public static PIDFTalonSRX MOTOR = new PIDFTalonSRX(
-                CAN.MOTOR_ID.INTAKE_OPENER_MOTOR_ID, INTAKE_OPENER_MOTOR_CONFIG);
+                CAN.IntakeOpener.MOTOR_ID, MOTOR_CONFIG);
+    }
+
+    protected static class PitcherComponents {
+        private static final MotorConfig MOTOR_CONFIG = new MotorConfig().
+                brake().
+                inverted(true).
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/RobotComponents.java:121]: defualt false/delete inverted

File: src/main/java/frc/robot/constants/RobotConstants.java [hunk 14]
```diff
@@ -117,8 +117,14 @@
...
         public static double OPENED_ANGLE = 97;
-        public static double CLOSED_ANGLE = 0;
         public static double STALL_CURRENT_LIMIT = 20;
     }
+
+    public static class PitcherConstants {
+        public static final TrigonTalonSRX MOTOR = PitcherComponents.MOTOR;
+        public static final double GEAR_RATIO = 10;
+        public static double OPEN_ANGLE = 50;
+        public static double CLOSED_ANGLE = 70;
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/RobotConstants.java:127]: add // in degrees beside both

File: src/main/java/frc/robot/utilities/EncoderConversions.java [hunk 15]
```diff
@@ -68,6 +70,7 @@ public static double MPSToFalcon(double velocity, double circumference, double g
 
     /**
      * @param magTicks the ticks from the mag sensor
+     * @param gearRatio the gear ratio between mag and Mechanism (set 1 as default)
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/EncoderConversions.java:73]: @ param gearRatio Gear Ratio between Mag and Mechanism (set to 1 for Mag RPM)

File: src/main/java/frc/robot/utilities/EncoderConversions.java [hunk 16]
```diff
@@ -1,5 +1,7 @@
 package frc.robot.utilities;
 
+import com.fasterxml.jackson.databind.ser.std.StdKeySerializers;
+
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/EncoderConversions.java:4]: remove

File: src/main/java/frc/robot/constants/RobotConstants.java [hunk 17]
```diff
@@ -123,6 +123,7 @@
     public static class PitcherConstants {
         public static final TrigonTalonSRX MOTOR = PitcherComponents.MOTOR;
         public static final double GEAR_RATIO = 10;
+        // in degrees beside both
         public static double OPEN_ANGLE = 50;
         public static double CLOSED_ANGLE = 70;
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/RobotConstants.java:128]: fix

**Review by Mboesn — CHANGES_REQUESTED** [review]

**Review by Mboesn — CHANGES_REQUESTED** [review]

**Pull request merged**